modules/currency-l1/src/main/scala/io/constellationnetwork/currency/l1/http/SwapRoutes.scala

```diff
@@ -0,0 +1,91 @@
+package io.constellationnetwork.currency.l1.http
+
+import cats.effect.Async
+import cats.effect.std.{Queue, Supervisor}
+import cats.syntax.all._
+
+import io.constellationnetwork.currency.l1.domain.error.InvalidSignature
+import io.constellationnetwork.currency.schema.currency.{CurrencyIncrementalSnapshot, CurrencySnapshotInfo}
+import io.constellationnetwork.currency.swap.ConsensusInput
+import io.constellationnetwork.error.ApplicationError._
+import io.constellationnetwork.ext.http4s.error._
+import io.constellationnetwork.node.shared.domain.cluster.storage.L0ClusterStorage
+import io.constellationnetwork.node.shared.domain.queue.ViewableQueue
+import io.constellationnetwork.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import io.constellationnetwork.routes.internal.{InternalUrlPrefix, _}
+import io.constellationnetwork.schema.swap.SwapTransaction
+import io.constellationnetwork.schema.{GlobalIncrementalSnapshot, GlobalSnapshotInfo}
+import io.constellationnetwork.security.signature.Signed
+import io.constellationnetwork.security.{Hasher, SecurityProvider}
+
+import eu.timepit.refined.auto._
+import io.circe.shapes._
+import org.http4s.circe.CirceEntityCodec.{circeEntityDecoder, circeEntityEncoder}
+import org.http4s.dsl.Http4sDsl
+import org.http4s.{HttpRoutes, Response}
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import shapeless._
+import shapeless.syntax.singleton._
+
+final case class SwapRoutes[F[_]: Async: Hasher: SecurityProvider](
+  swapPeerConsensusInput: Queue[F, Signed[ConsensusInput.PeerConsensusInput]],
+  l0ClusterStorage: L0ClusterStorage[F],
+  swapTransactionsQueue: ViewableQueue[F, Signed[SwapTransaction]],
+  lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+  lastCurrencySnapshotStorage: LastSnapshotStorage[F, CurrencyIncrementalSnapshot, CurrencySnapshotInfo]
+)(implicit S: Supervisor[F])
+    extends Http4sDsl[F]
+    with PublicRoutes[F]
+    with P2PRoutes[F] {
+
+  def logger = Slf4jLogger.getLogger[F]
+
+  protected val prefixPath: InternalUrlPrefix = "/"
+
+  private def validate(tx: Signed[SwapTransaction])(onValid: F[Response[F]]): F[Response[F]] =
+    // TODO: @mwadon - validation
+    onValid
+
+  protected val public: HttpRoutes[F] = HttpRoutes.of[F] {
+    case GET -> Root / "swap" =>
+      swapTransactionsQueue.view.flatMap(Ok(_))
+
+    case req @ POST -> Root / "swap" =>
+      req
+        .asR[Signed[SwapTransaction]] {
```

Comment:
Perhaps I am still misunderstanding but I thought the SwapTransaction was to be treated like Rewards and not generated by the end-user? Doesn't this structure require the user to sign and generate the swap transaction?

Reply:
Perhaps the `SwapTransaction` is what was previously called AllowSpend? In that case, I understand. I was thinking this is the balance updating transaction and not the "user authorization" portion of the process.

Reply:
Yes, it's a trait above the AllowSpend. @AlexBrandes wanted to make it possible to put new types of transactions in the future. Maybe SwapTransaction is a bit unfortunate name, we can think of something better.

Reply:
I think we should definitely refine and write down naming conventions for AMM 🤔

Reply:
yes, I was able to find it [here](https://github.com/Constellation-Labs/tessellation/blob/8dde3dd93cae65bb17d77fd7ca47e797962c8cac/modules/shared/src/main/scala/io/constellationnetwork/schema/swap.scala#L88). Thanks for clarifying as well.